crates/matrix-sdk-base/src/read_receipts.rs

```diff
@@ -0,0 +1,395 @@
+use matrix_sdk_common::deserialized_responses::SyncTimelineEvent;
+use ruma::{
+    events::{
+        poll::{start::PollStartEventContent, unstable_start::UnstablePollStartEventContent},
+        receipt::{ReceiptThread, ReceiptType},
+        room::message::Relation,
+        AnySyncMessageLikeEvent, AnySyncTimelineEvent, OriginalSyncMessageLikeEvent,
+        SyncMessageLikeEvent,
+    },
+    serde::Raw,
+    EventId, OwnedEventId, RoomId, UserId,
+};
+use tracing::{field::display, instrument, trace};
+
+use super::BaseClient;
+use crate::{error::Result, store::StateChanges, RoomInfo};
+
+/// Provider for timeline events prior to the current sync.
+pub trait PreviousEventsProvider: Send + Sync {
+    /// Returns the list of known timeline events, in sync order, for the given
+    /// room.
+    // TODO: return a reference or some kind of iterator
```

Comment:
Do you want to address that in this PR? If not, please open an issue and link it to the `TODO` note.

Reply:
I've done it in a later commit :eyes: